### ISSUE TITLE
Shellcode Extraction in Cobalt Strike Stagers

### ISSUE THREAD

**Issue by as0ni** [hasherezade/mal_unpack#6]
Hi,

First off, thank you for your fantastic contributions to the community! Your research, write-ups, and tools are incredibly inspiring.

I'm reaching out regarding an issue I encountered with mal_unpack while testing it against some Cobalt Strike stagers. Specifically, it appears that mal_unpack does not successfully extract shellcode in cases where the shellcode begins with the `CLD` (opcode `FC`) instruction.

For example, in this sample ([VT link](https://www.virustotal.com/gui/file/cde5c70c71c37dd3eb154a237e74afccc16486d5c2d9fc581938741f9451c222)), the shellcode is deobfuscated following a VirtualAlloc call and starts with `FC E8 89 00 00 00`. In a similar loader, I found that the shellcode began with the opcodes `FC 48 83 E4`. I also attempted using the `/shellc` option, but it did not identify the shellcode in these cases.

The recent addition of the `/pattern` option is helpful, and I was able to extract the shellcode by providing a custom sig file. However, I wanted to bring this to your attention as these shellcode patterns, starting with `FC`, are common in Cobalt Strike samples and might be worth considering as built-in signatures for improved detection.

Please note that I haven't tested this with the companion driver yet. If you think it might help in this case, I can give it a try and report back.

Thank you again for all your hard work on this project!

**Comment by hasherezade** [comment]
Hi @as0ni ! Thank you for your kind words, and for the report! Sure this will be useful to include in the built-in signatures. I will take a look, and update you soon.

**Comment by as0ni** [comment]
Thank you, @hasherezade! Also, if possible, could you point me to the code that contains the current built-in signatures? Thanks again.

**Comment by hasherezade** [comment]
@as0ni  - the default ones are here:  https://github.com/hasherezade/pe-sieve/blob/3f696a8610512db20a3854c988f9823c98b7edaa/utils/code_patterns.h

**Comment by hasherezade** [comment]
@as0ni  - sorry for the late response. I was busy with other things. But, looking closer I noticed that this pattern is already added to the default set. It is [this one](https://github.com/hasherezade/pe-sieve/blob/3f696a8610512db20a3854c988f9823c98b7edaa/utils/code_patterns.h#L49).

I was wondering what could be the reason of the issue you described, because I could not reproduce it on my side. Then I rewatched your video: https://youtu.be/D6Bm5vD78eY?t=405 - and I realized that the error is very simple - you are just missing the `/shellc` parameter. 

By default, mal_unpack detects only implanted PEs - that's why in the first case you presented, the payloads are dumped. But if you want to detect shellcodes, you need to enable it explicitly. Using the `/pattens` argument is an alternative way to enable the pattern check - but it will use the patterns from the set that you supplied. When you want to enable the built-in set, you simply add `/shellc P`.

The commandline that worked for me to unpack the shellcode from the [syswow.exe](https://github.com/as0ni/youtube-files/blob/main/syswow.zip) sample is:

```
mal_unpack.exe /shellc P /exe syswow.exe /timeout 10000 /trigger T
```

The option `/trigger T` is added so that the sample will not be killed immediately on the first shellcode detected, but instead keep running, and be killed on the timeout. That allows to dump all the payloads that are subsequently loaded.

Please have a look at the PE-sieve Wiki - it explains all the arguments and how they can be used: https://github.com/hasherezade/pe-sieve/wiki. Also, the detailed description of all the options is available using the argument `/help`.

**Comment by as0ni** [comment]
Hi @hasherezade, thank you for your detailed reply! During my initial testing, I did try the `/shellc` option but didn't realize that the sample is terminated after the first shellcode detection. After following your suggestion to use the `/trigger` option, I was able to successfully dump the code I was looking for. I'll add a comment to my YouTube video to include this clarification. 

I appreciate your guidance—thank you again.
